serve/mlc_serve/engine/engine_common.py

```diff
@@ -135,6 +137,62 @@ def detokenize_incrementally(
     return delta
 
 
+def logprob_detokenize(
+        tokenizer: TokenizerP,
+        logprob_info: Optional[RawLogprobsInfo],
+) -> Optional[LogprobsContent]:
+    """Detokenize tokens from RawLogprobInfo and convert the latter to LogprobContent"""
+    if logprob_info is None:
+        return None
+
+    top_logprobs: List[TopLogprobs] = []
+    if (
+        logprob_info.top_tokens is not None and
+        logprob_info.top_logprobs is not None
+    ):
+        top_tokens = list(zip(logprob_info.top_tokens, logprob_info.top_logprobs))
+        # dedup duplicates
+        # Todo: Make sure decode can generate different tokens
+        if logprob_info.previous_tokens is None:
+            logprob_info.previous_tokens = []
+        for top_token, top_logprob in top_tokens:
+            detokenized = tokenizer.convert_ids_to_tokens(logprob_info.previous_tokens + [top_token])[-1]
```

Comment:
detokenization with duplicates is not my code. And it will most likely even interfere with future loglikelyhood calculations due to it needs pure tokens. Yes, I plan to refactor it in next PR(s), using delta and hide dedup by some control key which is off by default.